### PR TITLE
[REA-45] Stream pausing

### PR DESCRIPTION
**Pull request by Jmgr**
This PR proposes an API to “pause” a stream, that is, deactivate its partitions. Publication to one of these partitions using the Liftbridge API “resumes” the stream, restarting its partitions.

## Context

Our use-case will include having a significant number of streams, but only a small fraction of those will be active at any given point in time (sparse streams).

In case of a server shutdown we would like to restart only active streams, not paused ones. The activity would record stream pausing and resuming, so that on server startup only the active streams would be restored.

The aims of this PR are twofold:

- inactive streams are paused so as to avoid the memory and CPU cost of maintaining stream activity and replication when there are no messages;

- by passing pause and resume events to a meta-event stream, we can schedule activity for a consumer pool more effectively.

This PR is therefore linked to the activity stream PR (https://github.com/liftbridge-io/liftbridge/pull/169).

## Implementation

The implementation proposed in this PR provides a new `PauseStream` API that allows clients to request that a stream be paused. Pausing a steam is very similar to closing a stream; that is, closing all of its partitions. Closing a partition, in turn, means that the server is no longer a leader or a follower for it and that its commit log is also closed. A “paused” flag is also set on the stream’s partitions. This allows us to spare CPU, memory usage and close unused file handles. One could also imagine an auto-pausing feature for each stream, where after a certain duration without any message publication a stream and its partition would be automatically paused. This would have to be added in a future PR however.

Resuming a stream occurs when a message is published to one of its partitions. All partition are then re-created using the same parameters, including the same data directory. The message is then published as usual.

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -167,6 +185,28 @@ func (a *apiServer) Publish(ctx context.Context, req *client.PublishRequest) (
 	}
 	a.logger.Debugf("api: Publish [subject=%s]", subject)
 
+	stream := a.metadata.GetStream(req.Stream)
+	if stream == nil {
+		return nil, status.Error(codes.NotFound, fmt.Sprintf("No such stream: %s", req.Stream))
+	}
+	partition, ok := stream.partitions[0] // Only check the first partition for now
```
Comment (paddybyers) [inline server/api.go]: So have we decided what the policy should be here? Do all partitions get paused simultaneously? Or they can independently pause and resume?

File: server/metadata.go [hunk 2]
```diff
@@ -29,8 +29,8 @@ const (
 // a stream partition that already exists.
 var ErrPartitionExists = errors.New("partition already exists")
 
-// ErrStreamNotFound is returned by DeleteStream when attempting to delete a
-// stream that does not exists.
+// ErrStreamNotFound is returned by DeleteStream/PauseStream when attempting to
+// delete/pause a stream that does not exists.
 var ErrStreamNotFound = errors.New("stream does not exists")
```
Comment (paddybyers) [inline server/metadata.go]: s/exists/exist/

Comment (Jmgr) [reply server/metadata.go]: Fixed in a3aabcd.

File: server/api.go (hunk 1 above)
Comment (Jmgr) [reply server/api.go]: We could add an option to perform either. Internally the stream does not store anything, so the pause status is stored into a partition struct. For now the API pauses the whole stream, that means all partitions. When one partition gets resumed, all other partitions from the stream also get resumed. I think we could see that Tyler says about this, unless you have some idea on what we should do?

Comment (paddybyers) [reply server/api.go]: ok

**Comment by Jmgr** [comment]
PR created upstream, so closing this.
See https://github.com/liftbridge-io/liftbridge/pull/170, https://github.com/liftbridge-io/liftbridge-api/pull/18, and https://github.com/liftbridge-io/go-liftbridge/pull/49.

**Pull request closed without merging**